src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -227,8 +361,502 @@ sealed abstract class AST(pos: Position, subexprs: Array[AST] = Array.empty) {
+  def memoize[T](mc: CM[Code[T]])(implicit tti: TypeInfo[T]): CM[(Code[Unit], Code[T])] = for (
+    c <- mc;
+    x <- newLocal[T]
+  ) yield (x.store(c), x.load().asInstanceOf[Code[T]])
+  def memoize[T](c: Code[T])(implicit tti: TypeInfo[T]): CM[(Code[Unit], Code[T])] = for (
+    x <- newLocal[T]
+  ) yield (x.store(c), x.load().asInstanceOf[Code[T]])
+
+  // references to its argument will be duplicated
+  def bindInRaw[T](name: String, c: Code[AnyRef])(body: CM[Code[T]]): CM[Code[T]] = CM { case (e, s) =>
+    body.mt(e.copy(m = e.m + ((name, c))), s)
+  }
+  def bindRepInRaw[T](bindings: Seq[(String, CM[Code[AnyRef]])])(body: CM[Code[T]]): CM[Code[T]] = bindings match {
+    case Seq() => body
+    case Seq((name, cmc), rest @ _*) =>
+      cmc.flatMap(c => bindInRaw(name, c)(bindRepInRaw(rest)(body)))
+  }
+  // its argument is stored in a local variable, only refernece to the variable
+  // are dupliacted
+  def bindIn[T](name: String, c: Code[AnyRef])(body: CM[Code[T]]): CM[Code[T]] =
+    memoize(c).flatMap { case (st, v) => bindInRaw(name, v)(body.map(x => Code(st, x))) }
+  def bindRepIn[T](bindings: Seq[(String, CM[Code[AnyRef]])])(body: CM[Code[T]]): CM[Code[T]] = bindings match {
+    case Seq() => body
+    case Seq((name, cmc), rest @ _*) =>
+      cmc.flatMap(c => bindIn(name, c)(bindRepIn(rest)(body)))
+  }
+  def lookup(name: String): CM[Code[AnyRef]] = CM { case (e, s) =>
+    (e.m(name), s)
+  }
+
+  def invokePrimitive1Pure[A,R](f: (A) => R)
+    (implicit act: ClassTag[A], rct: ClassTag[R]): CM[Code[A] => Code[R]] =
+    foreignFun(f).map((ff: Code[(A) => R]) => (a: Code[A]) => ff.invoke[A,R]("apply", a))
+
+  def invokePrimitive0[R](f: () => R)()(implicit rct: ClassTag[R]): CM[Code[R]] =
+    foreignFun(f).map(_.invoke[R]("apply"))
+  def invokePrimitive1[A,R](f: (A) => R)(a: Code[A])
+    (implicit act: ClassTag[A], rct: ClassTag[R]): CM[Code[R]] =
+    foreignFun(f).map(_.invoke[A,R]("apply", a))
+  def invokePrimitive2[A,B,R](f: (A, B) => R)(a: Code[A], b: Code[B])
+    (implicit act: ClassTag[A], bct: ClassTag[B], rct: ClassTag[R]): CM[Code[R]] =
+    foreignFun(f).map(_.invoke[A,B,R]("apply", a, b))
+  def invokePrimitive3[A,B,C,R](f: (A, B, C) => R)(a: Code[A], b: Code[B], c: Code[C])
+    (implicit act: ClassTag[A], bct: ClassTag[B], cct: ClassTag[C], rct: ClassTag[R]) : CM[Code[R]] =
+    foreignFun(f).map(_.invoke[A,B,C,R]("apply", a, b, c))
+  def invokePrimitive4[A,B,C,D,R](f: (A, B, C, D) => R)(a: Code[A], b: Code[B], c: Code[C], d: Code[D])
+    (implicit act: ClassTag[A], bct: ClassTag[B], cct: ClassTag[C], dct: ClassTag[D], rct: ClassTag[R]) : CM[Code[R]] =
+    foreignFun(f).map(_.invoke[R]("apply", Array[Class[_]](act.runtimeClass, bct.runtimeClass, cct.runtimeClass, dct.runtimeClass), Array(a, b, c, d)))
+  def createLambda[T <: AnyRef](name: String, body: CM[Code[T]])(implicit tti: TypeInfo[T]): CM[Code[(AnyRef) => T]] = {
```

Comment:
For when you need to pass a lambda as a `Function1[T, U]` argument to some Java/Scala code.